client/render/CBitmapHandler.cpp

```diff
@@ -103,21 +104,40 @@ SDL_Surface * BitmapHandler::loadH3PCX(ui8 * pcx, size_t size)
 	return ret;
 }
 
+ResourceID loadBitmapResource(const std::string & path, const std::string & fname)

```

Comment:
Technically we already have such functionality - see CMappedFileLoader class.
Was used in original version of wog mod, probably unused right now

Reply:
for some reason doesn't work.
If I add code below, other files from data folder become invisible
```
"filesystem" :
	{
		"DATA/" :
		[
			{"type" : "map", "path" : "Data/HeroScr4.pix", "initial" : "Data/HeroScr3.pix"},
		]
	}
```

Reply:
Is hero screen the only file that needs such replacement?
In that case I'd rather use "targeted" fix and all check to hero window.

Also, RoE (and maybe AB) don't have 5th MISC slot, which is probably why RoE has different hero screen image. Probably needs to be handled somehow?

Reply:
Ideally would be to rewrite heroWindow to make it configurable. But for now I'm trying to **launch** game, not adopt everything

Reply:
In that case maybe keep this as draft PR until demo support is properly integrated with minimal hacks in codebase?

Reply:
but why this is a hack and what "properly" means?

Reply:
Instead of "tricking" vcmi into thinking that data matches sod we should:

- have this integrated into base game. Meaning, vcmi should be usable with demo files as it, without any additional mods or any other manual configuration by player. 

- we should also either disable features added in AB/SoD or allow them to be properly used with RoE data. For example, MISC_5 slot should be either disabled when RoE data is used or visible in all UI elements (e.g. hero screen, exchange screen, etc).

- for things like hero screen background we should use some mechanism that will:
- 1) detect data type provided by player (roe demo, sod demo, full roe, full ab, full sod, full complete)
- 2) provide information specific to data type, for example - which file should be used for hero screen background, how many artifact slots are present, how many entries are in text files, etc.
- 3) this information can be either hardcoded or loaded from text config, but it should be in a base game and not in some mod that players have to find.
(this can be done for example similarly to how map format "feature levels" are done)

Reply:
I want to say here that I want to make resource handling aside of actual engine.
Ideally, game shouldn't know anything about presence of some picture. Everything should be done on loading side and hacks are allowed in mod itself.
We have similar functionality with jsons for animations, but not for images.

Reply:
> Ideally, game shouldn't know anything about presence of some picture

True, what we could do is to have something like this in our json's
```json
    "featureLevelROE" : { "heroScreenBackgroundName" : "HeroScr3.pcx", "hasMisc5Slot" : false },
    "featureLevelSOD" : { "heroScreenBackgroundName" : "HeroScr4.pcx", "hasMisc5Slot" : true }
```
And based on game data detection, vcmi should use either "featureLevelROE" or "featureLevelSOD" entry.

> We have similar functionality with jsons for animations, but not for images.

Thing is, I don't see need for such functionality for images. Reasoning:
1) In this case what we need is filesystem-abstraction level (or window configuration level), not image format level. Since we need to adapt to different data layouts, not to different format of image data
2) Reasoning behind json animations was not filesystem abstraction, but an alternative to binary h3 .def format. Since we needed to provide support for images sequences similar to one provided by .def files. Alternative to binary h3 .pcx images are bmp/png's, not json